libyul/optimiser/StackToMemoryMover.cpp

```diff
@@ -98,68 +117,160 @@ m_nameDispenser(_context.dispenser)
-			return m_memoryOffsetTracker(var.name);
-		});
-		if (!leftHandSideNeedsMoving)
-			return {};
-
-		langutil::SourceLocation loc = _stmt.location;
 
-		if (_variables.size() == 1)
+		langutil::SourceLocation  loc = _stmt.location;
+		if (_lhsVars.size() == 1)
 		{
-			optional<YulString> offset = m_memoryOffsetTracker(_variables.front().name);
-			yulAssert(offset, "");
-			return generateMemoryStore(
-				m_context.dialect,
-				loc,
-				*offset,
-				_stmt.value ? *std::move(_stmt.value) : Literal{loc, LiteralKind::Number, "0"_yulstring, {}}
-			);
+			optional<YulString> offset = m_memoryOffsetTracker(_lhsVars.front().name);
+			if (offset)
+				return generateMemoryStore(
+					m_context.dialect,
+					loc,
+					*offset,
+					_stmt.value ? *std::move(_stmt.value) : Literal{loc, LiteralKind::Number, "0"_yulstring, {}}
+				);
+			else
+				return {};
 		}
+		FunctionCall const* functionCall = get_if<FunctionCall>(_stmt.value.get());
+		yulAssert(functionCall, "");
+		auto rhsSlots = m_functionReturnVariables.at(functionCall->functionName.name) |
+			ranges::views::transform([&](auto const& _var) {
+				return m_memoryOffsetTracker(_var.name);
+			}) | ranges::to<vector<optional<YulString>>>;
+
+		// Nothing to do, if the right-hand-side remains entirely on the stack and
+		// none of the variables in the left-hand-side are moved.
+		if (
+			ranges::none_of(rhsSlots, [](optional<YulString> const& _slot) { return _slot.has_value(); }) &&
+			!util::contains_if(_lhsVars, [&](auto const& var) { return m_memoryOffsetTracker(var.name); })
+		)
+			return {};
 
-		VariableDeclaration tempDecl{loc, {}, std::move(_stmt.value)};
 		vector<Statement> memoryAssignments;
 		vector<Statement> variableAssignments;
-		for (auto& var: _variables)
+		VariableDeclaration tempDecl{loc, {}, std::move(_stmt.value)};
+
+		yulAssert(rhsSlots.size() == _lhsVars.size(), "");
+		for (auto&& [lhsVar, rhsSlot]: ranges::views::zip(_lhsVars, rhsSlots))
 		{
-			YulString tempVarName = m_nameDispenser.newName(var.name);
-			tempDecl.variables.emplace_back(TypedName{var.location, tempVarName, {}});
+			unique_ptr<Expression> rhs;
+			if (rhsSlot)
+				rhs = make_unique<Expression>(generateMemoryLoad(m_context.dialect, loc, *rhsSlot));
+			else
+			{
+				YulString tempVarName = m_nameDispenser.newName(lhsVar.name);
+				tempDecl.variables.emplace_back(TypedName{lhsVar.location, tempVarName, {}});
+				rhs = make_unique<Expression>(Identifier{loc, tempVarName});
+			}
 
-			if (optional<YulString> offset = m_memoryOffsetTracker(var.name))
+			if (optional<YulString> offset = m_memoryOffsetTracker(lhsVar.name))
 				memoryAssignments += generateMemoryStore(
 					m_context.dialect,
 					loc,
 					*offset,
-					Identifier{loc, tempVarName}
+					move(*rhs)
 				);
 			else
 				variableAssignments.emplace_back(StatementType{
-					loc, {move(var)},
-					make_unique<Expression>(Identifier{loc, tempVarName})
+					loc, { std::move(lhsVar) },
```

Comment:
```suggestion
					loc,
					{std::move(lhsVar)},
```

Reply:
Ok, now I'm confused - are std::move here and move above two different functions?

Reply:
I'll be removing any `std::` left in the file :-).